src/main.rs

```diff
@@ -491,6 +624,53 @@ impl From<CheckRelease> for cargo_semver_checks::Check {
     }
 }
 
+/// Helper function to encapsulate the logic of validating that unstable options
+/// were not used without `-Z unstable-options` and issuing deprecation warnings
+/// for any stable feature flags that were explicitly specified.
+fn validate_feature_flags(
+    config: &mut GlobalConfig,
+    unstable_options: &UnstableOptions,
+) -> anyhow::Result<()> {
+    // needed to avoid borrow checker errors when printing with config.
+    let stable_flags: Vec<_> = config
+        .feature_flags()
+        .iter()
+        .filter(|x| x.stable)
+        .copied()
+        .collect();
+
+    for stable_flag in stable_flags {
+        config
+            .shell_warn(format_args!(
+                "the feature flag {} has been stabilized and may be removed
+            from the list of feature flags in a future release.",
+                stable_flag.id
+            ))
+            .expect("printing failed");
+    }
+
+    if !config.feature_flag_enabled(FeatureFlag::UNSTABLE_OPTIONS) {
+        let non_default_options = unstable_options.non_default();
+
+        if !non_default_options.is_empty() {
+            let mut message = String::from(
+                "The following options are not supported without `-Z unstable-options`:\n",
```

Comment:
I think our current `anyhow::bail!()` messages try to be lowercase, to match cargo and rustc. I'm assuming this will get printed as `error: The following ...` right?

If so, consider:
```suggestion
                "the following options are not supported without `-Z unstable-options`:\n",
```